crucible/src/Lang/Crucible/Simulator/GlobalState.hs

```diff
@@ -31,184 +39,462 @@ import           Lang.Crucible.Simulator.RegMap
 
 -- | A map from global variables to their value.
---   The integer value is the number of symbolic branches
---   currently pending.  We use this primarily as a sanity check
---   to help find bugs where we fail to match up calls to
---   globalPushBranch with globalAbortBranch/globalMuxFn
 data SymGlobalState (sym :: Type) =
   GlobalState
-  { _globalPendingBranches :: Int
-  , globalMap             :: MapF.MapF GlobalVar (GlobalEntry sym)
-  , globalReferenceMap    :: MapF.MapF RefCell (RefCellContents sym)
+  { globalFrames :: !(GlobalFrames sym)
+    -- ^ The stack of updates to global memory, separated by branch
+    -- pushes. This field only contains values with ordinary crucible
+    -- types (i.e. not intrinsic), which do not have mux operations
+    -- that require branch push notifications.
+  , globalIntrinsics :: !(GlobalContents sym)
+    -- ^ The set of updates since initialization to vars and refs that
+    -- have intrinsic or "any" types, which must be notified of
+    -- branch pushes.
   }
 
--- | The empty set of global variable bindings
+-- | The empty set of global variable bindings.
 emptyGlobals :: SymGlobalState sym
-emptyGlobals = GlobalState 0 MapF.empty MapF.empty
+emptyGlobals = GlobalState emptyGlobalFrames emptyGlobalTable
+
+-- | Test whether this type could be an intrinsic type, which must be
+-- notified of branch pushes and aborts.
+needsNotification :: TypeRepr tp -> Bool
+needsNotification tr =
+  case tr of
+    IntrinsicRepr{} -> True
+    AnyRepr -> True
+    _ -> False
 
--- | Lookup a global in the state
+-- | Lookup a global variable in the state.
 lookupGlobal :: GlobalVar tp -> SymGlobalState sym -> Maybe (RegValue sym tp)
-lookupGlobal g gst = globalEntryValue <$> MapF.lookup g (globalMap gst)
+lookupGlobal g gst
+  | needsNotification (globalType g) =
+      globalEntryValue <$> MapF.lookup g (globalVars (globalIntrinsics gst))
+  | otherwise =
+      globalEntryValue <$> go (globalFrames gst)
+  where
+    go (InitialFrame c) = MapF.lookup g (globalVars c)
+    go (BranchFrame u c _) = MapF.lookup g (globalVars u) <|> MapF.lookup g (globalVars c)
 
 -- | Set the value of a global in the state.
-insertGlobal :: GlobalVar tp
-             -> RegValue sym tp
-             -> SymGlobalState sym
-             -> SymGlobalState sym
-insertGlobal g v gst = gst{ globalMap = MapF.insert g (GlobalEntry v) (globalMap gst) }
+insertGlobal ::
+  GlobalVar tp ->
+  RegValue sym tp ->
+  SymGlobalState sym ->
+  SymGlobalState sym
+insertGlobal g v gst
+  | needsNotification (globalType g) =
+      gst{ globalIntrinsics = updateGlobalVars (MapF.insert g x) (globalIntrinsics gst) }
+  | otherwise =
+      gst{ globalFrames = upd (globalFrames gst) }
+  where
+    x = GlobalEntry v
+    upd (InitialFrame c) = InitialFrame (updateGlobalVars (MapF.insert g x) c)
+    upd (BranchFrame u c gf) = BranchFrame (updateGlobalVars (MapF.insert g x) u) c gf
+    -- TODO: maybe the BranchFrame case should be an error?
```

Comment:
Why are you thinking this should be an error?

Reply:
I was thinking that creating a global variable inside a branch should be disallowed. But now I realize that `insertGlobal` is also used to update already-existing global vars, which is totally reasonable inside a branch. So I should delete that comment.

Reply:
Ahh, indeed...

Perhaps we should be more careful here and actually ensure that a global variable exists before simply writing to it, and generate an error immediately instead of waiting until a mux happens.